script/corpora_text.py

```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/env python3
+import numpy as np
+import sys
+import os
+from collections import OrderedDict
+import argparse
+
+
+class CorporaText:
+    def __init__(self, collection_name):
+        self.corpora = np.memmap(collection_name, dtype=np.uint32, mode='r')
+        doc_start = self.build_start_index()
+
+        with open(collection_name + '.documents', 'r') as f:
+            docnos = f.read().split()
+            self.documents = OrderedDict(zip(docnos, doc_start))
+
+        with open(collection_name + '.terms', 'r') as f:
+            self.terms = np.array(f.read().split())
+
+    def doc_text(self, docno):
+        if docno not in self.documents:
+            return []
+        start = self.documents[docno]
+        size = self.corpora[start]
+        seq = self.corpora[start + 1:start + 1 + size]
+        return self.terms[seq].tolist()
+
+    def build_start_index(self):
+        i = 2
+        starting = []
+        while i < len(self.corpora):
+            size = self.corpora[i]
+            starting.append(i)
+            i += size + 1
+        return starting
+
+    def __len__(self):
+        return self.corpora[1]
+
+    def __iter__(self):
+        for docno in self.documents:
+            yield (docno, self.doc_text(docno))
+
+    def __next__(self):
+        return self
+
+
+def file_path(prefix):
+    if not all(
+            os.path.isfile(p)
+            for p in [prefix, prefix + '.documents', prefix + '.terms']):
+        raise argparse.ArgumentTypeError("Please check the prefix is valid.")
+    return prefix
+
+
+def parse_arguments():
+    parser = argparse.ArgumentParser(
+        description='Dump document text.',
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+    parser.add_argument('-i',
+                        '--interactive',
+                        action='store_true',
+                        help='Read docno from stdin')
+
+    parser.add_argument('prefix', type=file_path)
+
+    return parser.parse_args()
+
+
+def main():
+    args = parse_arguments()
+    corpora = CorporaText(args.prefix)
```

Comment:
Can you clarify what is the input here? Forward index or inverted index?

Reply:
The input is docnos, the output is the content of the doc dumped from the forward index.

Reply:
Isn't this basically the same as this?

```
read_collection - read collections.
Usage: /home/michal/pisa/build/bin/read_collection [OPTIONS] first [last]

Positionals:
  first INT REQUIRED          Element number
  last INT                    Element number

Options:
  -h,--help                   Print this help message and exit
  -c,--collection TEXT REQUIRED
                              Collection file name
  -m,--map TEXT               String map file name
```

This, as far as I remember, works both for forward and inverted index. The advantage of it might be performance. Or does your program do something more I missed?

Reply:
Great to know. I think I missed this program. I'll try it.